Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs

```diff
@@ -0,0 +1,147 @@
+        /// If a filter string is set, MoveNext will advance to the next entry that matches the filter.
+        /// </summary>
+        /// <returns>
+        /// true if the enumerator was successfully advanced to the next element; false if the enumerator has passed the end of the collection
+        /// </returns>
+        public bool MoveNext()
+        {
+            IsCurrentValid = fileInfoEnumerator.MoveNext();
+            while (IsCurrentValid && IsCurrentHidden())
+            {
+                IsCurrentValid = fileInfoEnumerator.MoveNext();
+            }
+
+            return IsCurrentValid;
+        }
+
+        /// <summary>
+        /// Attempts to save the filter string for this enumeration.  When setting a filter string, if Current is valid
+        /// and does not match the specified filter, the enumerator will be advanced until an element is found that
+        /// matches the filter (or the end of the collection is reached).
+        /// </summary>
+        /// <param name="filter">Filter string to save.  Can be null.</param>
+        /// <returns> True if the filter string was saved.  False if the filter string was not saved (because a filter string
+        /// was previously saved).
+        /// </returns>
+        /// <remarks>
+        /// Per MSDN (https://msdn.microsoft.com/en-us/library/windows/hardware/ff567047(v=vs.85).aspx, the filter string
+        /// specified in the first call to ZwQueryDirectoryFile will be used for all subsequent calls for the handle (and
+        /// the string specified in subsequent calls should be ignored)
+        /// </remarks>
+        public bool TrySaveFilterString(
+            string filter)
+        {
+            if (filterString == null)
+            {
+                SaveFilter(filter);
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Returns the current filter string or null if no filter string has been saved
+        /// </summary>
+        /// <returns>The current filter string or null if no filter string has been saved</returns>
+        public string GetFilterString()
+        {
+            return filterString;
+        }
+
+        private static bool FileNameMatchesFilter(
+            string name,
+            string filter)
+        {
+            if (string.IsNullOrEmpty(filter))
+            {
+                return true;
+            }
+
+            if (filter == "*")
+            {
+                return true;
+            }
+
+            return Utils.IsFileNameMatch(name, filter);
+        }
+
+        private void SaveFilter(
+            string filter)
+        {
+            if (string.IsNullOrEmpty(filter))
+            {
+                filterString = string.Empty;
+            }
+            else
+            {
+                filterString = filter;
+                if (IsCurrentValid && IsCurrentHidden())

```

Comment:
>if [](start = 16, length = 2)

Why not loop until filter match?

Reply:
Not sure. This is how the sample chose to do this. I don't want to change behavior to risk breaking things.

---
In reply to: [298338318](https://github.com/microsoft/BuildXL/pull/519#discussion_r298338318) [](ancestors = 298338318)